tests/func/test_diff.py

```diff
@@ -163,17 +163,49 @@ def test_diff_no_cache(tmp_dir, scm, dvc):
         {"dir": {"file": "file content"}}, commit="first"
     )
     scm.tag("v1")
+
     tmp_dir.dvc_gen(
         {"dir": {"file": "modified file content"}}, commit="second"
     )
+    scm.tag("v2")
 
-    remove(first(stage.outs).cache_path)
-    remove("dir")
+    remove(dvc.cache.local.cache_dir)
 
     # invalidate_dir_info to force cache loading
     dvc.cache.local._dir_info = {}
 
-    diff = dvc.diff("v1")
+    diff = dvc.diff("v1", "v2")
     assert diff["added"] == []
     assert diff["deleted"] == []
     assert first(diff["modified"])["path"] == os.path.join("dir", "")
+
+
+def test_diff_dirty(tmp_dir, scm, dvc):
+    tmp_dir.dvc_gen(
+        {"file": "file_content", "dir": {"dir_file1": "dir file content"}},
+        commit="initial",
+    )
+
+    (tmp_dir / "file").unlink()
+    tmp_dir.gen({"dir": {"dir_file2": "dir file 2 content"}})
+    tmp_dir.dvc_gen("new_file", "new_file_content")
+
+    result = dvc.diff()
+
+    assert result == {
+        "added": [
+            {"hash": "86d049de17c76ac44cdcac146042ec9b", "path": "new_file"}
+        ],
+        "deleted": [
+            {"hash": "7f0b6bb0b7e951b7fd2b2a4a326297e1", "path": "file"}
```

Comment:
I have a problem with `deleted`, because user will not be able to commit stage file when there is no output. But I presumed that its better to just log `deleted` instead some ambiguous warning saying "Output is deleted but stage file still exists, so you will not be able to commit current state of `file`".

Reply:
`deleted` seems ok to me, since it's consistent with what `dvc status` reports for this case
```
file.dvc:
        changed outs:
                deleted:            file
```